src/tapedfunction.jl

```diff
@@ -239,97 +267,152 @@ end
-    bind_var!(Symbol(:_, var.id), bindings, ir.slottypes[var.id])
-bind_var!(var::Core.SlotNumber, bindings::Dict{Symbol, Any}, ir::Core.CodeInfo) =
-    bind_var!(Symbol(:_, var.id), bindings, ir.slottypes[var.id])
-bind_var!(var::Symbol, boxes::Dict{Symbol, Any}, c::Core.Const) =
-    bind_var!(var, boxes, _loose_type(Type{c.val}))
-bind_var!(var::Symbol, boxes::Dict{Symbol, Any}, c::Core.PartialStruct) =
-    bind_var!(var, boxes, _loose_type(c.typ))
-function bind_var!(var::Symbol, bindings::Dict{Symbol, Any}, ::Type{T}) where T
-    get!(bindings, var, nothing)
-    Box{T}(var)
+const IRVar = Union{Core.SSAValue, Core.SlotNumber}
+
+struct TempBindings
+    data::Bindings
+    book::Dict{IRVar, Int}
+end
+
+function bind_var!(var_literal, tbind::TempBindings, ir::Core.CodeInfo)
+    # for literal constants
+    push!(tbind.data, var_literal)
+    idx = length(tbind.data)
+    return idx
+end
+function bind_var!(var::GlobalRef, tbind::TempBindings, ir::Core.CodeInfo)
+    in(var.mod, (Base, Core)) ||
+        @info "evaluating GlobalRef $var at compile time"
+    bind_var!(getproperty(var.mod, var.name), tbind, ir)
+end
+function bind_var!(var::QuoteNode, tbind::TempBindings, ir::Core.CodeInfo)
+    @info "evaluating QuoteNode $var at compile time"
+    bind_var!(eval(var), tbind, ir)
+end
+function bind_var!(var::Core.TypedSlot, tbind::TempBindings, ir::Core.CodeInfo)
+    # turn TypedSlot to SlotNumber
+    bind_var!(Core.SlotNumber(var.id), tbind, ir)
+end
+function bind_var!(var::Core.SlotNumber, tbind::TempBindings, ir::Core.CodeInfo)
+    get!(tbind.book, var, allocate_binding!(var, tbind, ir.slottypes[var.id]))
+end
+function bind_var!(var::Core.SSAValue, tbind::TempBindings, ir::Core.CodeInfo)
+    get!(tbind.book, var, allocate_binding!(var, tbind, ir.ssavaluetypes[var.id]))
+end
+
+allocate_binding!(var, tbind::TempBindings, c::Core.Const) =
+    allocate_binding!(var, tbind, _loose_type(Type{c.val}))
+allocate_binding!(var, tbind::TempBindings, c::Core.PartialStruct) =
+    allocate_binding!(var, tbind, _loose_type(c.typ))
+function allocate_binding!(var, tbind::TempBindings, ::Type{T}) where T
+    # we may use the type info (T) here
+    push!(tbind.data, nothing)
+    idx = length(tbind.data)
+    return idx
 end
 
 function translate!(tape::RawTape, ir::Core.CodeInfo)
-    bindings = Dict{Symbol, Any}()
+    bindings = Bindings()
+    sizehint!(bindings, 128)
+    bcache = Dict{IRVar, Int}()
+    tbind = TempBindings(bindings, bcache)
+    slots = Dict{Int, Int}()
 
     for (idx, line) in enumerate(ir.code)
         isa(line, Core.Const) && (line = line.val) # unbox Core.Const
-        ins = translate!!(Core.SSAValue(idx), line, bindings, ir)
+        isconst = isa(ir.ssavaluetypes[idx], Core.Const)
+        ins = translate!!(Core.SSAValue(idx), line, tbind, isconst, ir)
         push!(tape, ins)
     end
-    return (bindings, tape)
+    for (k, v) in bcache
```

Comment:
Maybe rewrite for clarity 
```suggestion
    for (k, v) in tbind.book 
```